arangod/Scheduler/ThreadPoolScheduler.cpp

```diff
@@ -0,0 +1,123 @@
+void ThreadPoolScheduler::trackQueueItemSize(std::int64_t int64) noexcept {}
+
+uint64_t ThreadPoolScheduler::getLastLowPriorityDequeueTime() const noexcept {
+  return _lastLowPriorityDequeueTime.load(std::memory_order::relaxed);
+}
+
+void ThreadPoolScheduler::setLastLowPriorityDequeueTime(
+    uint64_t time) noexcept {
+  _lastLowPriorityDequeueTime.store(time, std::memory_order::relaxed);
+}
+
+std::pair<uint64_t, uint64_t>
+ThreadPoolScheduler::getNumberLowPrioOngoingAndQueued() const {
+  auto dequeued =
+      _threadPools[int(RequestPriority::LOW)]->statistics.dequeued.load();
+  auto queued =
+      _threadPools[int(RequestPriority::LOW)]->statistics.queued.load();
+  auto remaining = dequeued > queued ? 0 : queued - dequeued;
+  return std::make_pair(
+      _metrics->_ongoingLowPriorityGauge.load(std::memory_order_relaxed),
+      remaining);
+}
+
+double ThreadPoolScheduler::approximateQueueFillGrade() const { return 0; }
+
+double ThreadPoolScheduler::unavailabilityQueueFillGrade() const { return 0; }
+
+bool ThreadPoolScheduler::queueItem(RequestLane lane,
+                                    std::unique_ptr<WorkItemBase> item,
+                                    bool bounded) {
+  item->enqueueTime = std::chrono::steady_clock::now();
+  auto prio = PriorityRequestLane(lane);
+  _threadPools[int(prio)]->push(std::move(item));
+  return true;
+}
+
+ThreadPoolScheduler::ThreadPoolScheduler(
+    ArangodServer& server, uint64_t maxThreads,
+    std::shared_ptr<SchedulerMetrics> metrics)
+    : Scheduler(server), _metrics(std::move(metrics)) {
+  ThreadPoolMetrics poolMetrics = {
+      .jobsDone = &_metrics->_metricsJobsDoneTotal,
+      .jobsQueued = &_metrics->_metricsJobsSubmittedTotal,
+      .jobsDequeued = &_metrics->_metricsJobsDequeuedTotal,
+  };
+
+  _threadPools.reserve(4);
+
+  poolMetrics.queueLength = _metrics->_metricsQueueLengths[0];
+  poolMetrics.dequeueTimes = _metrics->_metricsDequeueTimes[0];
+  _threadPools.emplace_back(std::make_unique<WorkStealingThreadPool>(
+      "SchedMaintenance", std::max(std::ceil(maxThreads * 0.1), 2.),
+      poolMetrics));
+
+  poolMetrics.queueLength = _metrics->_metricsQueueLengths[1];
+  poolMetrics.dequeueTimes = _metrics->_metricsDequeueTimes[1];
+  _threadPools.emplace_back(std::make_unique<WorkStealingThreadPool>(
+      "SchedHigh", std::max(std::ceil(maxThreads * 0.4), 8.), poolMetrics));
+
+  poolMetrics.queueLength = _metrics->_metricsQueueLengths[2];
+  poolMetrics.dequeueTimes = _metrics->_metricsDequeueTimes[2];
+  _threadPools.emplace_back(std::make_unique<WorkStealingThreadPool>(
+      "SchedMedium", std::max(std::ceil(maxThreads * 0.4), 8.), poolMetrics));
+
+  poolMetrics.queueLength = _metrics->_metricsQueueLengths[3];
+  poolMetrics.dequeueTimes = _metrics->_metricsDequeueTimes[3];
+  _threadPools.emplace_back(std::make_unique<WorkStealingThreadPool>(
+      "SchedLow", std::max(std::ceil(maxThreads * 0.6), 16.), poolMetrics));
```

Comment:
The multipliers of `maxThreads` sum up to `0.6 + 0.4 + 0.4 + 0.1 == 1.5`. I would have expected them to sum up to just `1.0` here. That we use 1.5 times the number of configured threads may be intentional, but the rationale for it is not immediately clear. I suggest adding code comments here to explain the rationale.

Reply:
This has been done, because there is no longer thread sharing between the different priorities.

Reply:
I will add a comment. FWIW - in the future I want to get rid of the separate thread pools and have a WorkStealingThreadPool where we have separate partitions for the threads to steal from.